java/ext/javadoctest.py

```diff
@@ -0,0 +1,92 @@
+import os
+import pathlib
+import subprocess
+
+from sphinx.ext.doctest import (Any, Dict, DocTestBuilder, TestcodeDirective,
+                                TestoutputDirective, doctest, sphinx)
+from sphinx.locale import __
+
+
+class JavaDocTestBuilder(DocTestBuilder):
+    """
+    Runs java test snippets in the documentation.
+    """
+
+    name = "javadoctest"
+    epilog = __(
+        "Java testing of doctests in the sources finished, look at the "
+        "results in %(outdir)s/output.txt."
+    )
+
+    def compile(
+        self, code: str, name: str, type: str, flags: Any, dont_inherit: bool
+    ) -> Any:
+        # go to project that contains all your arrow maven dependencies
+        path_arrow_project = os.path.join(pathlib.Path.cwd(), "source", "demo")
```

Comment:
Should we use `tempfile. mkdtemp()` instead of messing with the source directory?

Reply:
To compile arrow code java need to localize [some jar libraries needed](https://github.com/apache/arrow-cookbook/blob/0d04b9684e0707b98b0144f3ad2f1543eca68336/java/source/demo/pom.xml#L17:L53), this demo project is a mockup project contains our library needed for our cookbook examples

Reply:
Also could be possible to create this demo project on the fly